lib/auth/webauthncli/fido2.go

```diff
@@ -562,54 +582,277 @@ func runOnFIDO2Devices(
+
+		for receiveCount < devices.len() {
+			select {
+			case <-devicesC:
+				receiveCount++
+			case <-maxWait.C:
+				log.Debugf("FIDO2: Abandoning device goroutines after %s", fido2DeviceMaxWait)
+				return
+			}
+		}
+		log.Debug("FIDO2: Device goroutines exited cleanly")
+	}()
+
+	// First "interactive" response wins.
+	for receiveCount < devices.len() {
+		select {
+		case err := <-devicesC:
+			receiveCount++
+
+			// Keep going on cancels or non-interactive errors.
+			if errors.Is(err, libfido2.ErrKeepaliveCancel) || errors.Is(err, &nonInteractiveError{}) {
+				log.Debugf("FIDO2: Got cancel or non-interactive device error: %v", err)
+				continue
+			}
+
+			return trace.Wrap(err)
+
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		}
+	}
+	return trace.Wrap(errors.New("all MFA devices failed"))
+}
+
+func startDevices(
+	locs []*libfido2.DeviceLocation,
+	filter deviceFilterFunc,
+	deviceCallback deviceCallbackFunc,
+	prompt runPrompt) (devices *openedDevices, devicesC <-chan error, err error) {
+	fidoDevs := make([]FIDODevice, 0, len(locs))
+	openDevs := make([]*openedDevice, 0, len(locs))
+
+	// closeAll should only be used until the devices are handed over.
+	// Do not defer-call it.
+	closeAll := func() {
+		for i, dev := range fidoDevs {
+			path := openDevs[i].path
+			err := dev.Close()
+			log.Debugf("FIDO2: Close device %v, err=%v", path, err)
+		}
+	}
+
+	// Open all devices in one go.
+	// This ensures cancels propagate to the complete list.
+	for _, loc := range locs {
+		path := loc.Path
+
+		dev, err := fidoNewDevice(path)
+		if err != nil {
+			closeAll()
+			return nil, nil, trace.Wrap(err, "device open")
+		}
+
+		fidoDevs = append(fidoDevs, dev)
+		openDevs = append(openDevs, &openedDevice{
+			path: path,
+			dev:  dev,
+		})
+	}
+
+	// Prompt touch, it's about to begin.
 	ackTouch, err := prompt.PromptTouch()
 	if err != nil {
+		closeAll()
+		return nil, nil, trace.Wrap(err)
+	}
+	//nolint:ineffassign // closeAll not meant to be used from here onwards.
+	closeAll = nil
+
+	errC := make(chan error, len(fidoDevs))
+	devices = &openedDevices{
+		devices: openDevs,
+	}
+
+	// Fire device handling goroutines.
+	// From this point onwards devices are owned by their respective goroutines,
+	// only cancels are supposed to happen outside of them.
+	for i, dev := range fidoDevs {
+		path := openDevs[i].path
+		dev := dev
+		go func() {
+			errC <- handleDevice(path, dev, filter, deviceCallback, devices.cancelAll, ackTouch, prompt)
+		}()
+	}
+
+	return devices, errC, nil
+}
+
+type openedDevice struct {
+	path string
+
+	// dev is the opened device.
+	// Only cancels may be issued outside of the handleDevice goroutine.
+	dev interface{ Cancel() error }
+
+	// Keep tabs on canceled devices to avoid multiple cancels.
+	canceled bool
+}
+
+type openedDevices struct {
+	mu      sync.Mutex // guards devices changes and cancelAll
+	devices []*openedDevice
+}
+
+func (l *openedDevices) len() int {
+	return len(l.devices)
```

Comment:
~~this should either acquire the lock or be named like `lenLocked` if the lock is supposed to be held already.~~

Nvm I see that the slice isn't supposed to change after `openedDevices` is created - can you clarify in the `mu` comment?

Reply:
Done. Happy to, given that this one got 2 separate comments.